addons/package_hierarchy/models/package_links.py

```diff
@@ -0,0 +1,182 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, models, fields, _
+from odoo.exceptions import ValidationError
+
+
+class PackageHierarchyLink(models.Model):
+    _name = "package.hierarchy.link"
+    _description = "Links for constructing package hierarchies"
+    _order = "id"
+
+    _parent_name = "parent_id"
+
+    name = fields.Char(index=True, store=True, compute="_compute_name")
+
+    parent_id = fields.Many2one(
+        "stock.quant.package", string="Parent package", ondelete="cascade", check_company=True
+    )
+    child_id = fields.Many2one(
+        "stock.quant.package", string="Child package", ondelete="cascade", check_company=True
+    )
+    move_line_ids = fields.Many2many(
+        "stock.move.line",
+        column1="link_id",
+        column2="move_line_id",
+        string="Stock move line",
+        ondelete="cascade",
+        check_company=True,
+    )
+    company_id = fields.Many2one("res.company", default=False)
+    depth = fields.Integer(compute="_compute_would_be_depth")
```

Comment:
Removing this field, I was originally planning on using it for checks on depth but it wasn't needed and the depth of a link is somewhat hard to define as it may be fluid.